### DISCUSSION TITLE
Python version not showing in prompt

### DISCUSSION THREAD

**Discussion by Griffin-Brome** [starship/starship#1287] [General]
Hello everyone,

I like to code in python so I want to be able to see my current python version when I'm working in a directory that contains .py files. However, starship does not display anything regarding python when in such a directory.

If someone could help me I would be very grateful. I am running starship in bash on Ubuntu 20.04.

Thank you!

```
 # Configuration for Starship bash prompt


# Don't print a new line at the start of the prompt
add_newline = false

# Replace the "❯" symbol in the prompt with "λ"
[character]      # The name of the module we are configuring is "character"
symbol = "λ"     # The "symbol" segment is being set to "➜"

# Disable the package module, hiding it from the prompt completely
[package]
disabled = true


# Make prompt 1 line
[line_break]
disabled = true

# Disable command duration prompt
[cmd_duration]
disabled = true

# Python settings
[python]
scan_for_pyfiles = true
```

Here is my config file.

**Comment by andytom** [comment]
Hi @Griffin-Brome I have a few questions to troubleshoot this. What is the output when you run `python --version`? Do you have an alias for `python`, e.g what does `alias python` output? Can you run `RUST_LOG=trace starship module python` from a directory you would expect python to be enabled and post the result here.

I think the problem is going to be related to Ubuntu 20.04 defaulting to python 3 only which is by default only installed as `python3` and starship tried to execute `python`.

**Reply by Griffin-Brome** [reply]
Hey @andytom I've run the commands you suggested. It looks like you are right, since I have no "python" alias. 
```
~ λ python --version

Command 'python' not found, did you mean:

  command 'python3' from deb python3
  command 'python' from deb python-is-python3

~ λ type python
bash: type: python: not found
~ λ alias python
bash: alias: python: not found
```

```
DEBUG starship::config > Config parsed: 
Table({"add_newline": Boolean(false), "character": Table({"symbol": String("λ")}), "package": Table({"disabled": Boolean(true)}), "line_break": Table({"disabled": Boolean(true)}), "cmd_duration": Table({"disabled": Boolean(true)}), "python": Table({"scan_for_pyfiles": Boolean(true)})})
 DEBUG starship::config > Config found for "python": 
Some(Table({"scan_for_pyfiles": Boolean(true)}))
 TRACE starship::context > Building HashSets of directory files, folders and extensions took 67.617µs
 TRACE starship::utils   > Executing command "python" with args ["--version"]
 TRACE starship::utils   > Executing command "python" failed by: Os { code: 2, kind: NotFound, message: "No such file or directory" }

```

**Reply by Griffin-Brome** [reply]
I added ```alias python='python3' ```to my .bashrc, but now when I'm inside a python project, I get a red "[?]" in my prompt like this: 
``` 
parrot on master [?] λ 
```

**Reply by andytom** [reply]
This is a bit of a complicated one that I don't think we can completely fix in Starship 😞.

To try to give full context the problem is that there is no `python` binary installed in Ubuntu 20.04 by default, only `python3`. When Starship detects some python files is it executes `python --version` to get the version of python that is installed, this as we can see from the logs, fails since there is no `python` binary.
```
 TRACE starship::utils   > Executing command "python" with args ["--version"]
 TRACE starship::utils   > Executing command "python" failed by: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```
This is the same as when you executed the `python --version` command, the fix is to add an alias, however Starship is unable to resolve the shell aliases, see #632, and it would be tricky to add. We could add a flag to the python module to switch the command it runs, e.g  `python --version` or `python3 --version`, but I'm not sure this is the best approach.

Another possible solution would be to install the `python-is-python3` package but that might cause issues for other programs or scripts that expect `python` to be Python 2, this is also not currently reccomended by the Python Foundation, see [PEP 394](https://legacy.python.org/dev/peps/pep-0394/).

**Reply by chipbuster** [reply]
@andytom Would it be an option to add a hidden (advanced) feature to the python module to specify what command should be run to find "python"? For example, `pythoncommand = "python3"` or `pythoncommand = "pypy2"`

While I'm not a huge fan of this idea, the fact that most Linuxes have goobered the handling of the Python transition means that this might be a very common problem in the near future. It's also not one that we can wait out, since the Python 3.8 release notes note that certain language features will be fundamentally incompatible with Python 4, and require another version break :(

While this isn't strictly orthogonal to our current design, I think I would have to (begrudgingly) sacrifice purity to practicality on this point.

**Reply by andytom** [reply]
I think that is the only thing we can do at the moment, I like your suggestion of using a command argument rather than a flag. I'll try to take a look at this later today and see if I can get something working.

**Comment by dev-rajeshsinha** [comment]
Try `python3 --version` or `py --version`

**Reply by Griffin-Brome** [reply]
here is the output of those commands 
```
~ λ python3 --version
Python 3.8.2
~ λ py --version

Command 'py' not found, but can be installed with:

sudo apt install pythonpy


```
